File: Repository.php
<?php

require_once 'JsonAccessTrait.php';

class Repository {
    use JsonAccessTrait;    // Трейт для операцій з json файлами
    // Константи для роботи з конфігом
    const CONFIG_PATH = 'settings/config_db.json';
    const CONFIG_DEFAULT = [
        "host"      =>      "localhost",
        "dbname"    =>      "webappdb",
        "user"      =>      "root",
        "pass"      =>      ""
    ];

    // Абстракція для операцій з базою даних з використанням PDO
    // basic CRUD (Create, Read, Update, Delete)
    private PDO $connection; // властивість, що містить об’єкт з’єднання PDO
    private string $tableName; // властивість, яка відстежує назву таблиці, над якою виконуватимуться операції CRUD
    private array $columns; // властивість, що містить масив усіх столбців бд з яким потрібно роботати

    /**
     * @param string $tableName
     * @param array $columns
     */
    public function __construct(string $tableName, array $columns) {
        // Завантажує параметри підключення до бази даних із конфігураційного файлу
        // $config = require 'settings/db_config.php';
        $config = self::getJsonConfig();

        $dsn = "mysql:host={$config['host']};dbname={$config['dbname']};charset=utf8";

        $options = [
            PDO::ATTR_ERRMODE            => PDO::ERRMODE_EXCEPTION,
            PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
            PDO::ATTR_EMULATE_PREPARES   => false,
        ];
        $this->connection = new PDO($dsn, $config['user'], $config['pass'], $options);

        // Дані о таблиці з якою потрібно буде роботати
        $this->checkTable($tableName);
        $this->tableName = $tableName;
        $this->columns = $columns;
    }

    public function addRow(array $values): int { // Create
        // Вставляє новий рядок у таблицю.
        if (count($this->columns) !== count($values))
            // Перевіряє, чи однакова кількість стовпців і значень
            // Якщо ні, винекне помилка
            throw new InvalidArgumentException("addRow: Кількість стовпців і значення не збігаються.");


        foreach ($values as &$value)
            if ($value instanceof DateTime) // Конвертує об'єкт DateTime у строковий формат
                $value = $value->format('Y-m-d H:i:s');
            elseif(is_bool($value)) // Конвертує t/f у формат, у якому вони зберігаються у БД
                if($value == true) $value = 1;
                else $value = 0;

        $columnString = implode("`, `", $this->columns);
        $placeholders = implode(", ", array_fill(0, count($values), "?"));

        $query = "INSERT INTO `$this->tableName` (`$columnString`) VALUES ($placeholders)";
        // echo $query;

        // Готує SQL-запит. Цей метод повертає об’єкт PDOStatement.
        // Готуючи оператор, налаштовується запит на виконання у спосіб, який захищений від впровадження (injection) SQL.
        $stmt = $this->connection->prepare($query);

        // Виконання підготовленого оператора з фактичними значеннями.
        // Якщо виконання успішне, повертає останній вставлений ідентифікатор (який зазвичай є первинним ключем рядка).
        // Інакше повернути 0.
        return $stmt->execute($values) ? $this->connection->lastInsertId() : 0;
    }

    private function checkTable(string $tableName): void { // Read-Create
        // Перевіряє, чи існує таблиця в базі даних
        $query = "SHOW TABLES LIKE " . $this->connection->quote($tableName);
        $stmt = $this->connection->prepare($query);
        $stmt->execute();
        if ($stmt->rowCount() > 0) {
            $this->tableName = $tableName;
        }
        else { // Якщо ні, то назва таблиці шукається у масиві заготовок, після чого створюється на основі заготовки
            $tables = require 'settings/table_templates.php';
            foreach ($tables as $table => $query)
                if ($table == $tableName) {
                    $this->connection->exec($query);
                    return;
                }
            // Якщо немає заготовки для таблиці, то виникає помилка
            throw new InvalidArgumentException("checkTable(string $tableName): Таблиці $tableName не існує. Ймовірно, вона була випадково видалена, створить нову, будь ласка.");
        }
    }

    public function getAll(): array { // Read
        // Отримує всі рядки з таблиці
        $stmt = $this->connection->query("SELECT * FROM $this->tableName");
        return $stmt->fetchAll();
    }
    
    public function getRow($element, $column='id'): array { // Read
        // Отримує окремий рядок з таблиці, знаходячи його по $column
        error_log("getRow($element, $column): array");

        if (!$this->isThereRow($column, $element))
            throw new InvalidArgumentException("getRow($element): Рядка з $column $element в таблиці $this->tableName не існує.");

        $stmt = $this->connection->prepare("SELECT * FROM $this->tableName WHERE $column = :element");
        $stmt->bindParam(":element", $element, PDO::PARAM_INT);
        $stmt->execute();

        return $stmt->fetch(PDO::FETCH_ASSOC) ?: [];
    }

    public function isThereRow(string $column, string $rowElement): bool { // Read
        // Перевіряє, чи існує рядок з заданим елементом у певной колонці
        $query = "SELECT * FROM `$this->tableName` WHERE `$column` = :element";
        $stmt = $this->connection->prepare($query);
        $stmt->bindParam(':element', $rowElement, PDO::PARAM_STR);
        $stmt->execute();

        return $stmt->rowCount() > 0;
    }

    public function updateRow(int $id, array $values): bool { // Update
        // Оновлює певний рядок, визначений ідентифікатором
        if (count($this->columns) !== count($values))
            // Перевіряє, чи однакова кількість стовпців і значень
            // Якщо ні, винекне помилка
            throw new InvalidArgumentException("updateRow: Кількість стовпців і значення не збігаються.");

        if ($id < 1)
            throw new InvalidArgumentException("updateRow: Замовлення з id $id не існує.");

        foreach ($values as &$value)
            if ($value instanceof DateTime) // Конвертує об'єкт DateTime у строковий формат
                $value = $value->format('Y-m-d H:i:s');
            elseif(is_bool($value)) // Конвертує t/f у формат, у якому вони зберігаються у БД
                if($value == true) $value = 1;
                else $value = 0;

        $set = [];
        foreach ($this->columns as $index => $column)
            $set[] = "`$column` = ?";

        $setString = implode(", ", $set);

        $query = "UPDATE `$this->tableName` SET $setString WHERE id = ?";

        // Готує SQL-запит. Цей метод повертає об’єкт PDOStatement.
        $stmt = $this->connection->prepare($query);

        // Додає ідентифікатор до кінця значень для WHERE умови
        $values[] = $id;

        // Виконання підготовленого оператора з фактичними значеннями.
        // Повертає true у разі успіху, інакше false.
        return $stmt->execute($values);
    }

    public function removeRow(int $id): int { // Delete
        // Видаляє певний рядок, визначений ідентифікатором.
        $query = "DELETE FROM `$this->tableName` WHERE id = ?";
        $stmt = $this->connection->prepare($query);
        $stmt->execute([$id]);
        return $stmt->rowCount();
    }

    public static function createDB(string $dbname): void {
        // Метод для створення бд
        $config = self::getJsonConfig();

        try {
            // Створення підключення PDO
            $pdo = new PDO("mysql:host={$config['host']}", $config['user'], $config['pass']);
            // Встановлення режиму помилок PDO на виключення
            $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

            // SQL-запит для створення нової БД
            $sql = "CREATE DATABASE $dbname";

            // Виконання SQL-запиту
            $pdo->exec($sql);
        } catch(PDOException $e) {
             echo "Помилка при створенні бази даних: " . $e->getMessage();
        }

        $pdo = null; // Закриття підключення
    }

    public static function checkDB(string $dbname): bool {
        // Перевірка, чи існує база даних
        $config = self::getJsonConfig();

        try {
            $pdo = new PDO("mysql:host={$config['host']}", $config['user'], $config['pass']);
            $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

            $stmt = $pdo->query("SELECT COUNT(*) FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = '$dbname'");

            if ($stmt->fetchColumn() > 0)
                return true;
            else return false;

        } catch(PDOException $e) {
            echo "Помилка підключення: " . $e->getMessage();
            return false;
        }
    }

}

File: scripts/Repository.py
import mysql.connector
import json
import os


class Repository:
    config_path = os.path.join(os.path.dirname(__file__), '..', 'settings', 'config_db.json')

    def __init__(self, table):
        # Отримання параметрів підключення
        with open(self.config_path, 'r') as file:
            data = json.load(file)
        # Встановлення параметрів підключення
        config = {
            'host': data['host'],
            'user': data['user'],
            'password': data['pass'],
            'database': data['dbname']
        }

        self.table = table  # Таблиця, з якою будуть проводитися маніпуляції

        # Підключення до бази даних
        self.conn = self.connect(config)

    def connect(self, config):
        # Створення з'єднання
        try:
            connect = mysql.connector.connect(**config)
            if connect.is_connected():
                print(f'Connected to database ({self.table})')
                return connect

        except mysql.connector.Error as err:
            print(f'Error: {err}')

    def close_connect(self):
        # Закриття з'єднання
        if self.conn.is_connected():
            self.conn.close()
            print(f'The connection is closed ({self.table})')

    def get_all(self):
        # Повертає усі рядки бд
        cursor = self.conn.cursor()
        try:
            query = f'SELECT * FROM {self.table}'  # SQL запит для отримання рядків
            cursor.execute(query)
            rows = cursor.fetchall()  # Виконання SQL запиту
            return rows
        except mysql.connector.Error as err:
            print(f'Помилка запиту: {err}')
            return False
        finally:
            cursor.close()

    def get_all_in_dic(self):
        # Повертає усі рядки бази даних у якості словнику, де ключ є id
        rows = self.get_all()
        dic = {}
        for row in rows:
            dic[row[0]] = row
        return dic

    def get_row(self, value, column='id'):
        # Метод для отримання змісту рядку з бд

        cursor = self.conn.cursor()
        try:
            query = f'SELECT * FROM {self.table} WHERE {column} = %s'  # SQL запит для отримання рядка
            cursor.execute(query,
                           (value,))  # встановлення value в SQL запит через метод execute (захист від SQL ін'єкції)
            row = cursor.fetchone()  # Виконання SQL запиту
            return row
        except mysql.connector.Error as err:
            print(f'Помилка запиту: {err}')
            return False
        finally:
            cursor.close()

    def create_row(self, data):
        # Метод, для додавання рядку в бд

        columns = ', '.join([f'`{column}`' for column in data.keys()])  # Отримання колонок, в які треба додати данні
        placeholders = ', '.join(['%s' for _ in data.values()])  # Отримання даних, які треба додати в колонки

        query = f'INSERT INTO `{self.table}` ({columns}) VALUES ({placeholders})'  # SQL запит для створення рядку

        cursor = self.conn.cursor()

        try:
            cursor.execute(query, tuple(data.values()))  # Виконання SQL запиту
            self.conn.commit()  # Збереження результатів
            return True
        except mysql.connector.Error as err:
            print(f'Помилка запиту: {err}')
            return False
        finally:
            cursor.close()
